Mathlib/Analysis/Convex/ConvexFunction.lean

```diff
@@ -0,0 +1,541 @@
+  have a_nonneg : 0 ≤ a := by
+    have h1: 0 + b ≤ a + b := by
+      rw[zero_add, sum_a_b]
+      exact min_le_right b1 1
+    rw[add_le_add_iff_right b] at h1
+    exact h1
+  specialize cxf a_nonneg b_nonneg sum_a_b
+  let x' := a • x + b • y
+  have x'rfl : x' = a • x + b • y := rfl
+  have h1 : ‖x - x'‖ = ‖b • (x - y)‖ := by
+    congr; apply point_proportion a_nonneg b_nonneg sum_a_b x'rfl
+  have h2 : ‖b • (x - y)‖ = b * ‖x - y‖ := by
+    rw[norm_smul]
+    rw[Real.norm_of_nonneg]
+    apply b_nonneg
+  have x1nbhd: ‖x - x'‖ ≤ δ := by
+    rw[h1, h2]
+    have h3: b * ‖x - y‖ ≤ b1 * ‖x - y‖:= by
+      rw[mul_le_mul_right]
+      apply min_le_left
+      exact h₃
+    have h4: b1 * ‖x - y‖ = δ := by
+      rw[div_mul_cancel]
+      apply ne_of_gt h₃
+    rw[← h4]
+    apply h3
+  specialize converge x' x1nbhd
+  have H1: f x + (f' x) (x' - x) - e1 * ‖x - x'‖ ≤ f x':= by
+    have l1: f x + (f' x) (x' - x) - f x' ≤ ‖f x' - f x - (f' x) (x' - x)‖ := by
+      rw[Real.norm_eq_abs]
+      have : f x + (f' x) (x' - x) - f x' = -(f x' - f x - (f' x) (x' - x)) := by
+        ring
+      rw [this]
+      apply neg_le_abs_self
+    have l2: f x + (f' x) (x' - x) - f x'≤ e1 * ‖x - x'‖:= by
+      apply le_trans l1 converge
+    linarith
+  have H2: f x' ≤ a • f x + b • f y := by apply cxf
+  have H3: f y = f x + (f' x) (y - x) - ε := by simp only [map_sub, sub_sub_cancel]
+  have l3: a • f x + b • f y= a * (f x) + b * (f y) := by exact rfl
+  have l4: e1 * ‖x - x'‖ = ε * b / 2 := by
+    rw[h1, h2]
+    calc
+      e1 * (b * ‖x - y‖) = ε / (2 * ‖x - y‖) * (b * ‖x - y‖):= by rfl
+      _ = ((ε / 2) / ‖x - y‖) *(b * ‖x - y‖):= by ring
+      _ = ((ε / 2) / ‖x - y‖) * ‖x - y‖ * b := by rw[mul_comm b, mul_assoc]
+      _ = (ε / 2) * b := by rw [div_mul_cancel]; apply ne_of_gt h₃
+      _ = ε * b / 2 := by ring
+  rw [l4] at H1; rw [l3] at H2
+  have H4: a * f x + b * f y = f x + b * (f' x) (y - x) - b * ε := by rw [H3]; ring
+  have l5: b* (f' x) (y - x) = (f' x) (x' - x):= by
+    have h5: (x' - x) = b • (y - x)  := by
+      calc
+        x' - x = -(x - x'):= Eq.symm (neg_sub x x')
+        _ = - (b • (x - y)):= by rw [point_proportion a_nonneg b_nonneg sum_a_b x'rfl]
+        _ = -(b • x - b • y):= by rw[smul_sub]
+        _ = b • y - b • x:= by simp only[neg_sub]
+        _ = b • (y - x):= by rw[smul_sub]
+    rw [h5, lnmp b]
+  rw [l5] at H4
+  rw [H4] at H2
+  have H6: f x + (f' x) (x' - x) - ε * b / 2 ≤ f x + (f' x) (x' - x) - b * ε := le_trans H1 H2
+  have H7: - ε * b / 2 ≤ - b * ε := by linarith
+  have H8: - ε * b / 2 + b * ε ≤ 0 := by linarith
+  have blt: 0 < b:= by apply lt_min; apply b1pos; apply zero_lt_one
+  have H10: ε * b ≤ 0:= by linarith
+  have H11: ε ≤ 0:= nonpos_of_mul_nonpos_left H10 blt
+  linarith
+
+theorem first_order_condition_inverse
+    (h : ∀ x ∈ s, HasFDerivAt f (f' x) x) (h₁: Convex ℝ s)
+    (h₂: ∀ (x : E), x ∈ s → ∀ (y : E), y ∈ s → f x + f' x (y - x) ≤ f y): ConvexOn ℝ s f := by
```

Comment:
When writing predicates inline, you can shorten them by using the same notational trick as in `h` in the line above:
```suggestion
    (h₂: ∀ x ∈ s, ∀ y ∈ s, f x + f' x (y - x) ≤ f y): ConvexOn ℝ s f := by
```